### PR TITLE
introduce notations

### PR DESCRIPTION
**Pull request by Vonuras**
* Pure Scientific -> if the exponent is bigger than 1e6 it will be displayed as "E1.111e8"
* Pure Engineering -> same as Scientific but only shows multiples of 3 as exponent
* Toggle like the Themes within settings
* refactored some other functions within format

### CODE REVIEW

**Comment by Vonuras** [comment]
I will actually revise this, once some of the other Pull Request are merged, as there are a lot of approved PRs, that are not merged yet, and some of them are interconnected to these changes

File: src/Synergism.ts [hunk 1]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
     player.dayTimer = (60 * 60 * 24 - (s + 60 * m + 60 * 60 * h))
 }
 
+const FormatList = ['', '', 'M', 'B', 'T', 'Qa', 'Qt', 'Sx', 'Sp', 'Oc', 'No', 'Dc', 'UDc', 'DDc', 'TDc', 'QaDc', 'QtDc', 'SxDc', 'SpDc', 'OcDc', 'NoDc', 'Vg', 'UVg', 'DVg', 'TVg', 'QaVg', 'QtVg', 'SxVg', 'SpVg', 'OcVg', 'NoVg', 'Tg', 'UTg', 'DTg', 'TTg', 'QaTg', 'QtTg', 'SxTg', 'SpTg', 'OTg', 'NTg', 'Qd', 'UQd', 'DQd', 'TQd', 'QaQd', 'QtQd', 'SxQd', 'SpQd', 'OcQd', 'NoQd', 'Qi', 'UQi', 'DQi', 'TQi', 'QaQi', 'QtQi', 'SxQi', 'SpQi', 'OQi', 'NQi', 'Se', 'USe', 'DSe', 'TSe', 'QaSe', 'QtSe', 'SxSe', 'SpSe', 'OcSe', 'NoSe', 'St', 'USt', 'DSt', 'TSt', 'QaSt', 'QtSt', 'SxSt', 'SpSt', 'OcSt', 'NoSt', 'Ocg', 'UOcg', 'DOcg', 'TOcg', 'QaOcg', 'QtOcg', 'SxOcg', 'SpOcg', 'OcOcg', 'NoOcg', 'Nono', 'UNono', 'DNono', 'TNono', 'QaNono', 'QtNono', 'SxNono', 'SpNono', 'OcNono', 'NoNono', 'Ce'];
+const FormatListShort = ['', 'K', 'M', 'B', 'T', 'Qa'];
+function letter(power: number,str: string[]) {
```
Comment (KhafraDev) [inline src/Synergism.ts:1833]:
```suggestion
function letter(power: number, str: string[]) {
```

File: src/Synergism.ts [hunk 2]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
...
+    const len = str.length;
+    function lN(n: number) {
+        let result = 1;
+        for (let j = 0; j < n; ++j) {
+            result = len*result+1;
```
Comment (KhafraDev) [inline src/Synergism.ts:1838]:
```suggestion
            result = len * result + 1;
```

File: src/Synergism.ts [hunk 3]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
...
+        }
+        return result;
+    }
+    if (power <= 5) {
+        return str[0];
+    }
+    power = Math.floor(power / 3);
+    let i=0;
+    while (power >= lN(++i)){}
```
Comment (KhafraDev) [inline src/Synergism.ts:1847]:
```suggestion
    while (power >= lN(++i));
```

File: src/Synergism.ts [hunk 4]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
...
+        for (let j = 0; j < n; ++j) {
+            result = len*result+1;
+        }
+        return result;
+    }
+    if (power <= 5) {
+        return str[0];
+    }
+    power = Math.floor(power / 3);
+    let i=0;
```
Comment (KhafraDev) [inline src/Synergism.ts:1846]:
```suggestion
    let i = 0;
```

File: src/Synergism.ts [hunk 5]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
...
+    while (power >= lN(++i)){}
+    if (i==1) {
+        return str[power-1];
+    }
+    power -= lN(i-1);
+    let ret = '';
+    while (i>0) {
+        ret += str[Math.floor(power/Math.pow(len,--i))%len];
```
Comment (KhafraDev) [inline src/Synergism.ts:1854]:
```suggestion
    while (i > 0) {
        ret += str[Math.floor(power/Math.pow(len,--i)) % len];
```

File: src/Synergism.ts [hunk 6]
```diff
@@ -1980,15 +2005,17 @@ export const format = (
         return !back
             ? frontFormatted
             : `${frontFormatted}${dec}${back}`;
-    } else if (power < 1e6) {
+    }
+    if (power < 1e6) {
```
Comment (KhafraDev) [inline src/Synergism.ts:2009]:
```suggestion
    } else if (power < 1e6) {
```

**Review by KhafraDev — CHANGES_REQUESTED** [review]
Looks good so far, just a few style things need to be addressed

File: src/Synergism.ts [hunk 7]
```diff
@@ -1934,26 +1951,34 @@ export const format = (
     if (power < -15) {
         return '0';
     }
-
+    if (player.notation == 'Pure Engineering') {
+        const powerOver = (power % 3 < 0) ? (3 + power % 3) : (power % 3);
+        power = power - powerOver;
+        mantissa = mantissa * Math.pow(10, powerOver)
+    }
+    if (player.notation == 'Pure Scientific' || player.notation == 'Pure Engineering') {
```
[the remark this answers is not in the document]
Comment (Vonuras) [reply src/Synergism.ts:1972]: This actually needs to be just an if, as there is a case, where you want the first if and this one to trigger

[the remark this answers is not in the document]
Comment (Vonuras) [reply src/Synergism.ts:1972]: No I know
There is an if at line 1967 (first if), and if you change the if at 1971/1972 (second if) to be an else if, you would not run the code which is inside both ifs, while having notation be 'Pure Scientific'

File: src/Synergism.ts [hunk 8]
```diff
@@ -1827,6 +1828,34 @@ const loadSynergy = async () => {
...
+        }
+        return result;
+    }
+    if (power <= 5) {
+        return str[0];
+    }
+    power = Math.floor(power / 3);
+    let i = 0;
+    while (power >= lN(++i)) ;
+    if (i == 1) {
```
Comment (KhafraDev) [inline src/Synergism.ts:1848]:
```suggestion
    if (i === 1) {
```

File: src/Synergism.ts [hunk 9]
```diff
@@ -1980,15 +2005,17 @@ export const format = (
...
         // If the power is less than 1e6 then apply standard scientific notation
         // Makes mantissa be rounded down to 2 decimal places
         const mantissaLook = (Math.floor(mantissa * 100) / 100).toLocaleString(undefined, locOpts);
         // Makes the power group 3 with commas
         const powerLook = padEvery(power.toString());
         // returns format (1.23e456,789)
         return `${mantissaLook}e${powerLook}`;
-    } else if (power >= 1e6) {
+    }
+    if (power >= 1e6) {
```
[the remark this answers is not in the document]
Comment (KhafraDev) [reply src/Synergism.ts:2030]: this doesn't seem to be resolved

File: src/Synergism.ts (hunk 8 above)
Comment (Vonuras) [reply src/Synergism.ts]: done

**Review by KhafraDev — APPROVED** [review]

**Comment by KhafraDev** [comment]
CI is failing because of one of my suggestions, sorry about that

File: src/Synergism.ts [hunk 10]
```diff
@@ -1827,9 +1828,37 @@ const loadSynergy = async () => {
...
+    while (i > 0) {
+        ret += str[Math.floor(power/Math.pow(len,--i)) % len];
+    }
+    return ret;
+}
+
 // Bad browsers (like Safari) only recently implemented this.
 //
-const supportsFormatToParts = typeof (Intl.NumberFormat.prototype as Intl.NumberFormat).formatToParts === 'function';
+const supportsFormatToParts = typeof (Intl.NumberFormat.prototype ).formatToParts === 'function';
```
Comment (KhafraDev) [inline src/Synergism.ts:1861]: this needs to be reverted

**Pull request merged**